### PR TITLE
Add ALL the SPHINCS

### LINKED ISSUE #144: Add Round 2 signature scheme SPHINCS+

### PR DESCRIPTION
**Pull request by joostrijneveld**
This adds remaining 35 variants of SPHINCS+.

#### Manually checked properties

* [ ] No stringification macros
* [ ] Output-parameter pointers in functions are on the left
* [ ] Negative return values on failure of API functions (within restrictions of FO transform).
* [ ] `const` arguments are labeled as `const`
* [ ] variable declarations at the beginning (except in `for (size_t i=...`)
* Optional:
  * [ ] All integer types are of fixed size, using `stdint.h` types (including `uint8_t` instead of `unsigned char`)
  * [ ] Integers used for indexing are of size `size_t`

Closes #144

### CODE REVIEW

**Comment by dstebila** [comment]
Will you be adding files to `test/duplicate_consistency` to track consistency requirements?  Or are all these unique files?

**Comment by thomwiggers** [comment]
When building one of the variants with:

```
CC=clang make clean-scheme functest SCHEME=sphincs-sha256-256s-simple TYPE=sign EXTRAFLAGS="-g -fsanitize=memory"
```

I get the following output

```
==5365==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x565104af6628 in PQCLEAN_SPHINCSSHA256256SSIMPLE_CLEAN_treehash /home/thom/git/phd/PQClean/PQClean/crypto_sign/sphincs-sha256-256s-simple/clean/utils.c:122:13                                                                                       
    #1 0x565104af5e51 in PQCLEAN_SPHINCSSHA256256SSIMPLE_CLEAN_treehash_FORS_HEIGHT /home/thom/git/phd/PQClean/PQClean/crypto_sign/sphincs-sha256-256s-simple/clean/utils.c:170:5                                                                            
    #2 0x565104af7b06 in PQCLEAN_SPHINCSSHA256256SSIMPLE_CLEAN_fors_sign /home/thom/git/phd/PQClean/PQClean/crypto_sign/sphincs-sha256-256s-simple/clean/fors.c:100:9                                                                                        
    #3 0x565104aecf91 in PQCLEAN_SPHINCSSHA256256SSIMPLE_CLEAN_crypto_sign_signature /home/thom/git/phd/PQClean/PQClean/crypto_sign/sphincs-sha256-256s-simple/clean/sign.c:172:5                                                                            
    #4 0x565104aee2bf in PQCLEAN_SPHINCSSHA256256SSIMPLE_CLEAN_crypto_sign /home/thom/git/phd/PQClean/PQClean/crypto_sign/sphincs-sha256-256s-simple/clean/sign.c:308:5                                                                                      
    #5 0x565104a81970 in test_sign /home/thom/git/phd/PQClean/PQClean/test/crypto_sign/functest.c:113:9
    #6 0x565104a812da in main /home/thom/git/phd/PQClean/PQClean/test/crypto_sign/functest.c:247:15
    #7 0x7fe921ff8222 in __libc_start_main (/usr/lib/libc.so.6+0x24222)
    #8 0x565104a0507d in _start (/home/thom/git/phd/PQClean/PQClean/bin/functest_sphincs-sha256-256s-simple_clean+0x1f07d)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/thom/git/phd/PQClean/PQClean/crypto_sign/sphincs-sha256-256s-simple/clean/utils.c:122:13 in PQCLEAN_SPHINCSSHA256256SSIMPLE_CLEAN_treehash                                                        
Exiting
```

**Comment by thomwiggers** [comment]
The issue has been identified: the sizes of the secret keys were not updated.

**Comment by thomwiggers** [comment]
SPHINCS breaks the maximum run time of the build services... I've started running the tests locally on a server here, but I still expect it to take a little while. Running the builds in the docker containers will take even longer...

**Comment by dstebila** [comment]
> SPHINCS breaks the maximum run time of the build services... I've started running the tests locally on a server here, but I still expect it to take a little while. Running the builds in the docker containers will take even longer...

If necessary we could split tests across KEMs and signatures.  If I counted correctly, there will be 80 (IND-CCA) KEMs and 85 signature schemes in round 2.

**Comment by thomwiggers** [comment]
Running the SPHINCS+ schemes with -O3 brought each individual test under 10 minutes, but we still exceed the 50-minute job timeout on Travis. On CircleCI, the `-fsanitize` build of `sphincs-haraka-192s-robust` exceeded the 10-minute timeout.

**Comment by dstebila** [comment]
> Running the SPHINCS+ schemes with -O3 brought each individual test under 10 minutes, but we still exceed the 50-minute job timeout on Travis. On CircleCI, the `-fsanitize` build of `sphincs-haraka-192s-robust` exceeded the 10-minute timeout.
> 
> On Sandor (i7-4770K CPU @ 3.50GHz), running `test_functests` without sanitizers already took 1.5h.

Yikes. `test_functests` does 3 * NTESTS = 3 * 5 = 15 signatures per implementation.  We could bring NTESTS down to 2 or 1 to try avoiding timeouts on CI.  Maybe we want to make this a parameter / environment variable, so that we can have a non-cloud-based integration test do more iterations once a week or so.

**Comment by thomwiggers** [comment]
> Yikes. `test_functests` does 3 * NTESTS = 3 * 5 = 15 signatures per implementation. We could bring NTESTS down to 2 or 1 to try avoiding timeouts on CI. Maybe we want to make this a parameter / environment variable, so that we can have a non-cloud-based integration test do more iterations once a week or so.

I've done this as part of #117

**Comment by thomwiggers** [comment]
I've also experimented locally with adding `-march=native` and timed the compile and run times of `functest_sphincs-haraka-256s-robust` in the QEMU-arm docker container were quite similar.  For AMD64 (i.e. my `Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz`) it also did not make a difference (which kind of surprised me...).

**Comment by thomwiggers** [comment]
`-flto` also does not seem to make a difference, at least on my laptop.

File: crypto_sign/sphincs-haraka-256f-simple/clean/haraka.c [hunk 1]
```diff
@@ -9,133 +14,678 @@ Plain C implementation of the Haraka256 and Haraka512 permutations.
...
+    int i;
+    for (i = 0; i < 4; i++) {
+        out[2 * i] = br_dec32le(in + 4 * i);
+        out[2 * i + 1] = br_dec32le(in + 4 * i + 16);
+    }
+    br_aes_ct_ortho(out);
 }
 
-void PQCLEAN_SPHINCSHARAKA256FSIMPLE_CLEAN_tweak_constants(
+void PQCLEAN_SPHINCSHARAKA128FSIMPLE_CLEAN_tweak_constants(
```
Comment (thomwiggers) [inline crypto_sign/sphincs-haraka-256f-simple/clean/haraka.c]: 256 -> 128?

Comment (joostrijneveld) [reply crypto_sign/sphincs-haraka-256f-simple/clean/haraka.c]: Good catch!

**Comment by joostrijneveld** [comment]
I've increased the 30 minute Travis timeout from #154 to 60, which is sufficient for the 36min that all the SPHINCS together currently take.

**Comment by joostrijneveld** [comment]
This currently suffers from two issues, neither of which I know how to resolve.

AppVeyor indicates that for SHA2 and Haraka, `test_compile_lib`, `test_functest` and `test_testvectors` fail, without any other error messages (other than a random `[00:39:30] Failure: AssertionError (Got unexpected return code 2) ... FAILED` halfway in).

Furthermore, AppVeyor has a maximum runtime of an hour, and is quite a bit slower than e.g. Travis. With 2/3rds of the costly tests failing, it still does not finish before the time runs out.

**Comment by joostrijneveld** [comment]
Ah, I forgot the crucial part of that comment: can someone with easy access to a Windows machine/VM try to run one of those failing tests to see if there's something AppVeyor is not telling us?

**Comment by joostrijneveld** [comment]
For completeness: currently failing CirleCI tests are due to a timeout. I don't have a solution right now.

In combination with #117 merging this won't actually destroy CI when integrating other schemes, but it may be a problem for the scheduled full tests and when all tests are re-ran for framework changes.

File: .circleci/config.yml [hunk 2]
```diff
@@ -17,7 +17,7 @@ version: 2.1
...
+          docker run -e CI=true -e PQCLEAN_ONLY_DIFF=1 -e PQCLEAN_SKIP_SCHEMES=sphincs-haraka-128f-robust,sphincs-haraka-192s-robust,sphincs-sha256-128f-robust,sphincs-sha256-192s-robust,sphincs-shake256-128f-robust,sphincs-shake256-192s-robust,sphincs-haraka-128f-simple,sphincs-haraka-192s-simple,sphincs-sha256-128f-simple,sphincs-sha256-192s-simple,sphincs-shake256-128f-simple,sphincs-shake256-192s-simple,sphincs-haraka-128s-robust,sphincs-haraka-256f-robust,sphincs-sha256-128s-robust,sphincs-sha256-256f-robust,sphincs-shake256-128s-robust,sphincs-shake256-256f-robust,sphincs-haraka-128s-simple,sphincs-haraka-256f-simple,sphincs-sha256-128s-simple,sphincs-sha256-256f-simple,sphincs-shake256-128s-simple,sphincs-shake256-256f-simple,sphincs-haraka-192f-robust,sphincs-haraka-256s-robust,sphincs-sha256-192f-robust,sphincs-sha256-256s-robust,sphincs-shake256-192f-robust,sphincs-shake256-256s-robust,sphincs-haraka-192f-simple,sphincs-haraka-256s-simple,sphincs-sha256-192f-simple,sphincs-sha256-256s-simple,sphincs-shake256-192f-simple,sphincs-shake256-256s-simple --rm -v `pwd`:`pwd` -w `pwd` "pqclean/ci-container:$ARCH" /bin/bash -c "
```
Comment (dstebila) [inline .circleci/config.yml]: In retrospect it would have been good if I had made wildcard matching in helpers.py, but oh well.

Comment (joostrijneveld) [reply .circleci/config.yml]: I briefly considered adding it, but then shrugged and copy-pasted the list of parameter sets ;)

**Comment by dstebila** [comment]
What's the status on this PR?

**Comment by joostrijneveld** [comment]
I think it's ready to be merged.

**Pull request merged**